### ISSUE TITLE
Abandon V8?

### ISSUE THREAD

**Issue by JerrySievert** [plv8/plv8js#364]
The question has come up as to whether we should abandon V8 due to the many issues of dealing with V8 being a project dependency:

* difficult to build
* huge build dependency (3-5gb on Linux, 2gb on MacOS)
* difficulty supporting windows
* lack of ABI compatibility
* package maintainers refuse to support V8

So, I started looking at what it would take to remove V8 completely.

The long and the short of it is that most of the code is a mix of:

* glue code connecting c++ v8 to c postgres and vice-versa
* security checks
* functions

Let's discuss what would be gained or lost by moving away from V8 and into another javascript engine.

**Comment by JerrySievert** [comment]
Note that the 2.3 branch would be the last officially released branch with v8, 3.0alpha would remain as-is, and any bug fixes or regressions would be back-ported as per current policy.

**Comment by jsommr** [comment]
What's your opinion on SpiderMonkey/libmozjs? It seems there's Debian [[1]](https://packages.debian.org/buster/libmozjs-60-0) and Ubuntu [[1]](https://packages.ubuntu.com/disco/libmozjs-60-0) [[2]](https://packages.ubuntu.com/bionic/libmozjs-52-0) packages available. I would love for Azure and AWS to keep supporting your extension, and using a lesser known engine might discourage them because of security concerns. I know nothing about how they evaluate what plugins they want to support though and how they ensure their servers are secured, so I might be totally wrong here. But I also believe SpiderMonkey is going to be maintained and optimized in many years. I like that there's an established corporation behind it, just like [Hermes](https://hermesengine.dev/) by Facebook could be worth a shot.

**Comment by JerrySievert** [comment]
I'm open to suggestions.  My goal right now is to figure out what the right direction will be.

In the mean time, I'm doing some PoC work decoupling v8, and determining MVP.  The plv8 source code is fairly entangled with v8 at this point, when it really doesn't need to be.

**Comment by pkit** [comment]
@JerrySievert 
I think V8 should stay just because it's practical, ubiquitous and gets security bugs fixed.
I would say that maybe riding on top of nodejs instead of v8 directly could be a good compromise?

**Comment by JerrySievert** [comment]
@pkit I'm having difficulty figuring out what you mean when you say "riding on top of nodejs".

**Comment by claytongulick** [comment]
I'm no C guru, but after @JerrySievert brought this up I took a look at the QuickJS source code, and I have to say, it's one of the cleanest most beautiful C projects I've ever seen. It reads like a novel, and looks brain-dead simple to build.

The only argument I can think of in favor of sticking with v8 is that it's probably the most well-tested, bullet-proof source code on the planet right now from a security standpoint. Which was the justification that Cloudflare used to launch their cloud function product based on v8 isolates.

I guess I'm not sure, since plv8/pljs doesn't execute untrusted client source code, if that's the most compelling thing though?

Regardless of the engine that's used moving forward, I think it's a good idea to change the name to pljs so that if there is a decision at some point to use a different engine, that's not a user-facing change - and also makes it consistent with plpython etc... with the language, not the engine in the name.

**Comment by pkit** [comment]
@JerrySievert use nodejs instead of v8
Like here https://github.com/laverdet/isolated-vm

@claytongulick when code can crash the whole database it is "untrusted" in the sense that even having bugs there may cost you a lot.

> Which was the justification that Cloudflare used to launch their cloud function product based on v8 isolates.

Unfortunately current v8 is not secure. https://bugs.chromium.org/p/v8/issues/detail?id=9368

**Comment by JerrySievert** [comment]
hm. I don't think that isolated-vm solves the package maintainer issue, nor would really count as being a trusted language extension without quite a few modifications - at which point any gain would be lost imo.

as for v8 being "secure".  yeah, it's definitely not, but unfortunately even core pg has crash bugs the mix.  v8's better than it was, post 6.4, when the spectre/meltdown mitigations went into affect, but there are still earlier versions in the wild.

**Comment by claytongulick** [comment]
@pkit I guess it depends on what you mean by "secure". Good enough to run untrusted remote code in one of the most widely distributed and used pieces of software on earth? Seems to fit my definition of secure, but opinions vary I guess. 

From https://blog.cloudflare.com/cloud-computing-without-containers/

> The only reason this was possible at all is the open-source nature of V8, and its standing as perhaps the most well security tested piece of software on earth.

**Comment by pkit** [comment]
> hm. I don't think that isolated-vm solves the package maintainer issue, nor would really count as being a trusted language extension without quite a few modifications - at which point any gain would be lost imo.

I probably meant to use isolated vm approach to the build process.
AFAIK they build it as a part of `npm install` process, sidestepping the package maintaining problem...
 
> as for v8 being "secure". yeah, it's definitely not, but unfortunately even core pg has crash bugs the mix. v8's better than it was, post 6.4, when the spectre/meltdown mitigations went into affect, but there are still earlier versions in the wild.

it's just used everywhere which should make the bugs/mitigations at least widely known

**Comment by pkit** [comment]
> From https://blog.cloudflare.com/cloud-computing-without-containers/

I have read the cloudfoundry PR back then, thanks :)
But they do not allow any untrusted parties there IIRC
It's still not good enough for a free or freemium platform.

**Comment by JerrySievert** [comment]
Bugs are widely known, yes, but many providers do not upgrade to newer plv8 versions (patches yes, but since v8’s abi isn’t backward compatible and they don’t backport to older abi’s, that is a problem).

**Comment by pkit** [comment]
By "many providers" you mean Amazon Aurora and friends? Looks like pain in the ass, for sure.
For my case no isolates is a no go anyway. Dunno what I can do here. Maybe ripping the isolate-related stuff from V8 into a small standalone codebase, like we did with NaCL, but I'm afraid it's a pretty hefty task. Preliminary research gives me hundreds of functions that are used in plv8->d8 bridge.

**Comment by claytongulick** [comment]
> I probably meant to use isolated vm approach to the build process.
> AFAIK they build it as a part of npm install process, sidestepping the package maintaining problem..

If I'm understanding @pkit properly, the proposal would be to use the nodejs packages being distributed already, like isolated-vm? It looks like isolated-vm links dynamically to node/v8 via the mechanism that node has built-in for addons, builds with node-gyp.

>hm. I don't think that isolated-vm solves the package maintainer issue, nor would really count as being a trusted language extension without quite a few modifications - at which point any gain would be lost imo.

Building like a node addon lets you ```#include <v8.h>``` using their prebuilt shared v8 lib, not <node.h> - wouldn't require code mod that way, would it? Or am I misunderstanding?

Since there's such a massive overlap between Node devs and plv8, I don't think it would be burdensome to folks to ```npm install plv8``` as a distribution mechanism instead of ```apt install``` or whatever.

Would that interfere with AWS / Google / Azure including plv8 in their default available extensions?

For a local build on windows, I'd prefer to do a ```npm install``` rather than the current instructions, (which I couldn't make work at all).

**Comment by JerrySievert** [comment]
Slow to respond. At a conference the rest of the week/weekend but I will try to address when I’m back on a computer.

**Comment by devinus** [comment]
First of all, I appreciate all the work that's been put into plv8.

With that said, I'd recommend finding new maintainers for plv8 instead and working on pljs as a separate project.

All of the issues you cited sound like burnout and I can understand that. But there's a reason plv8 is used and that's because quite simply it's by far the fastest and most ubiquitous JS engine available today.

Almost everybody using plv8 today is willing to make the speed/memory and build complexity/quality of using V8 over another engine.

I guarantee that if you abandoned V8 a fork would immediately appear.

**Comment by JerrySievert** [comment]
wow, that's uh, quite the statement there.

a solid refusal by package maintainers to deal with the v8 distribution model across the board ... burnout.  yeah, ok.

editing to add:

> Almost everybody using plv8 today is willing to make the speed/memory and build complexity/quality of using V8 over another engine.

this conversation started because yet another large platform, heroku, is abandoning plv8 because of package maintainers not willing to deal with the v8 distribution model - so the "almost everyone" argument holds no water, since this removes plv8 from tens of thousands of users.  this is a big problem, and will only get worse.

here's the official announcement from two days ago: https://devcenter.heroku.com/changelog-items/1677

the conversations about this have been ongoing both publicly and privately for much longer than two days.

**Comment by JerrySievert** [comment]
@pkit rds/aurora only make extensions that are trusted available.  amazon deals with upstream packages, but at least has the staff/ability to do maintenance.  my concern would be putting them in a situation where they would eliminate plv8 for either too difficult to maintain or too easy to break out.  requiring specific versions of an unrelated package to be installed in order to run opens a pretty big vector.

as for node-isolate specifically, these issues make me take pause:

> This project works best on nodejs LTS version 8.11.2 (or later) or nodejs current version 10.x (any version). It will compile and generally run alright on nodejs version 8.6.0 and later, but under load you will run into crashes and hangs due to some bugs in v8. 

> Please note that versions of nodejs 10.2.0 - 10.9.0 may crash while using the snapshot feature. Additionally, on nodejs 8.x if you try to create a snapshot that contains an instance of ArrayBuffer or TypedArray [i.e. Uint8Array, Float32Array, and so on] the process will crash. It will work fine on nodejs 10.0.0 and higher.

but, it's difficult for me to test - there are ABI issues compiling against `8.9.4`, so I updated to `8.16.1`, which gave different errors compiling against the ABI.

having crash issues against a large version swath of node, and a different version swath of v8 is an issue, having compilation problems against supported node versions is another issue - neither of these seem to solve the maintainability issues for package distributors.

@claytongulick having the ability to `#include <v8.h>` vs `#include <node.h>` doesn't do what you think it does.  node does not build a *shared* version of v8, it builds an embedded version of v8.  building a node add-on builds a dynamic library that is loaded into the node runtime, extending the execution where node is the primary application being executed.  in the case of plv8, Postgres is the primary application that loads dynamic libraries into it, passing execution to the library for short periods of time, where plv8 sandboxes v8 interactions.

for plv8 to deal with a situation like that, it would become a "broker" passing requests back and forth between postgres and very specific versions of node - I don't see how this could continue to a) be a trusted extension (with trusted being the postgres version of trusted), or b) solve the maintainability issue for package managers.  if anything, it seems like it would be even more maintenance given the ABI incompatibilities and the requirement to have a full node build environment on any host that needs to use plv8.  it seems like that would be going in the wrong direction.

**Comment by devinus** [comment]
@JerrySievert I didn't mean to be offensive, as burnout is nothing to be ashamed of as an open source maintainer.

I guess the point I was trying to make is that it's likely there still exists a not insignificant chunk of users who would pick plv8 over pljs, legacy production code using plv8, and all the downstream tooling and package ecosystem that already exists for plv8.

People using plv8 in production depend on their stored procedures/triggers/etc to behave in a predictable enough way to when were when they were developed. Replacing v8 also means degrading the performance characteristics and inevitably running into the subtle JS timing differences.

Under those circumstances it's safe to assume that a fork is inevitable. There will still be a need for there to be a package *somewhere* called plv8 that continues to get *at least* security updates despite the package being a complete PITA to maintain.

I wasn't trying to be disrespectful, it's just that with those assumptions it makes a lot more sense to start a new package. Maybe consider abstracting out the generic glue for both packages to use for some collaboration even if afterwords you want to leave a note that plv8 is abandonware pending a new maintainer.

I can also imagine a pljs actually being a framework with multiple pluggable engines and leaving the build pipeline to build out `pljs/v8` as `plv8` to the `pljs/v8` maintainer. In this scenario you get to keep all the upstream benefits and it doesn't seem like a "hostile takeover" to downstream consumers. Yes, usage of plv8 might shink or disappear entirely as downstream users lag or discontinue support because of build difficulties, lack of support, or pljs offering a more attractive alternative. Until that happens there exists a clear need for `plv8` to continue to exist in some form, so I think letting that process happen naturally is a better alternative.

TL;DR -- Nobody using plv8 in production today is going to be able to just simply swap out plv8 to pljs without serious regression testing, profiling/benchmarking, and rewriting code as it's not a 1:1 transfer and will be a breaking change. Switching to a different engine is effectively a deprecation notice to every user of plv8 today until they have time to do that, rewrite, and in some cases be forced to rethink their use of procedures entirely.

**Comment by JerrySievert** [comment]
>  I didn't mean to be offensive, as burnout is nothing to be ashamed of as an open source maintainer.

I didn't say you did, but you made an assumption that appears to be based on your misunderstanding of what the issues are, and made some very pointed statements.

> I guess the point I was trying to make is that it's likely there still exists a not insignificant chunk of users who would pick plv8 over pljs, legacy production code using plv8, and all the downstream tooling and package ecosystem that already exists for plv8.

there are multiple types of plv8 usage:

* users of hosted database services such as Heroku, RDS, or Azure - these are the largest overall number of users that are affected by any changes, ranging from v8 engine version, to jsonb conversion methods, to platforms officially supported by plv8, to package maintainers choices for support
* hosting providers, such as @pkit who use plv8 to offer services to customers
* large deployments in corporations, which is why the direct jsonb conversion code was written
* small deployments being used throughout self-hosted databases
* hobbyist
* package maintainers for operating systems, that feed 95% of the above

each of the decisions that have been made over the last several years about the direction, releases, bug fixes, and overall maintenance of plv8 have taken all of these into consideration before any decision or release has been made.  this is a given, and has not and will not change.

from your unedited response, because I feel that some of what you wrote merits a response:

> People using plv8 in production depend on their stored procedures/triggers/etc to behave in a predictable way they were when they were developed. 

unfortunately, that cannot apply when changing plv8 major versions, as plv8 policy has been that minor version changes (2.1 to 2.2, 2.2 to 2.3, etc) are related to v8 engine changes (of which all performance characteristics are thrown out the window), and major version changes (2.3 to 3.0) feature not only changes to the v8 engine (from 1.x to 2.x, and 2.x to 3.x a fairly large change), but also other big changes under the hood.  some of those changes end up being back-ported as compile-time feature flags.

patch versions are just that - patches that relate to regressions, bug fixes, postgres internal changes, and operating system changes.  on the last two, postgres internals changes have occurred in the last two postgres releases, and v8 compilation has broken on every macOS version release due to how v8 has managed its dependency checking.  these changes have been back ported up to 3 years, and across major versions.

trying to tie any sort of performance characteristics to a version of plv8 is really not advised, unless it is at a patch level, since v8 itself has drastically different performance characteristics across its versions.

> Replacing v8 also means degrading the performance characteristics and inevitably running into the subtle JS timing differences.

this depends entirely on usage patterns: interestingly because v8 itself has such a high startup penalty, the usage pattern of connect, execute, disconnect is actually improved under most circumstances; the usage pattern of a very long-running stored procedure, or the usage pattern of a long connection with the same role could end up being a performance decrease, but even that would depend entirely on how much data is flowing between v8 and postgres.  v8 has a high performance cost when moving between c++ and javascript compared to many other engines.  given that plv8 is often-times performing as a broker between the two environments even an interpreter that is running slower could end up with higher performance overall.

> Under those circumstances it's safe to assume that a fork is inevitable. There will still be a need for there to be a package somewhere called plv8 that continues to get at least security updates despite the package being a complete PITA to maintain.

first, I will point to another part of this discussion that was explicitly stated:

> > Note that the 2.3 branch would be the last officially released branch with v8, 3.0alpha would remain as-is, and any bug fixes or regressions would be back-ported as per current policy.

second, you seem to be conflating package maintenance and plv8 maintenance - this discussion comes out of the former, not the latter.  plv8 maintenance is the continued development and upkeep of plv8 as it pertains to features and bug fixes.  package maintenance is distribution of plv8 by operating system vendors and database vendors.  package maintenance is where the problem lies - it started when Debian maintainers stopped releasing because of the issues of v8 dependency, followed shortly by Ubuntu (who tends to use Debian packages), then by Fedora, which feeds into RedHat, and finally by the core Postgres package maintainers themselves - all because of the difficulties listed in the first comment of this thread in maintaining v8 dependencies.  this cascades into a mix of developers, end-users, and hosting providers abandoning the platform (as Heroku has recently done).

you can see official statements from package maintainer by reading https://github.com/plv8/plv8/issues/308, which this discussion was borne of.  there are a large number of these discussions in past tickets, including links to discussions, notifications, and tickets detailing every aspect of the issues.  there are also a large number of private email chains saying the same things, that I'm happy to forward if you don't fully grasp what the issue is.

I will note that this particular discussion came to a head when a core postgres committer found out that Heroku was eliminating plv8 and contacted both the google v8 team and me privately.

> I wasn't trying to be disrespectful, it's just that with those assumptions it makes a lot more sense to start a new package. Maybe consider abstracting out the generic glue for both packages to use for some collaboration even if afterwords you want to leave a note that plv8 is abandonware pending a new maintainer.

and you just made a bunch of assumptions - a) that a new package was never an option, b) that there's been no thought to abstractions and what the plv8 code itself contains, c) that plv8 would somehow become abandoned.

> I can also imagine a pljs actually being a framework with multiple pluggable engines and leaving the build pipeline to build out pljs/v8 as plv8 to the pljs/v8 maintainer. 

if you spent time looking at the lack of commonality between javascript engines, you likely would not suggest this - there is less commonality between javascript engines than between different languages:

* some engines are stack-based, others are heap based
* some are object oriented, some are functional, some are procedural

this leaves the only real commonality between them being that they all can execute some javascript and respond in some way - this means that to be able to support more than one the glue code would need to be exceedingly complicated and likely very slow, and as noted above passing through the c++/javascript membrane in v8 is already pretty slow compared to some other engines.


hopefully this wall of text answers enough of your comment, but feel free to respond if there's something I missed or that you would like clarification on.

these issues have weighed heavily on plv8 development over the last couple of years due to changes in upstream distribution.  this ticket being opened for community discussion is the next step in trying to figure out the best course of action for plv8, and any decision will take all aspects into account.

**Comment by JerrySievert** [comment]
@claytongulick your turn :)

> I'm no C guru, but after @JerrySievert brought this up I took a look at the QuickJS source code, and I have to say, it's one of the cleanest most beautiful C projects I've ever seen. It reads like a novel, and looks brain-dead simple to build.

between opening this issue 15 days ago and now, I've managed to squeeze in (between work that gets me paid, and a fantastic conference as well) multiple proofs of concept (actually MVP's) and am happy to comment on QuickJS specifically.

QuickJS's code is very clean, yes, but currently it does not cleanly compile under all plv8 supported platforms out of the box, nor directly support all of the features that plv8 currently supports (even ignoring isolates at the moment).  in order to be a full replacement, the code would need to be forked until (if?) the upstream supported all of the platforms/use cases.  this would be similar to what node.js did with v8, where it still maintains its own build system and patches.

I've managed to work around most of those issues, but it would still not be a direct replacement without additional work.

I'm happy to comment on other engines as well, but I'm going to stop short of putting together a full matrix until after this whole discussion has run its course.

**Comment by pkit** [comment]
You can safely ignore my usage of the platform. Because of the service
nature of my stuff I need to fork and maintain everything anyway.

**Comment by JerrySievert** [comment]
@pkit I appreciate that, but you still have a vested interest in this conversation and have provided both valuable insight and great code.

**Comment by devinus** [comment]
> trying to tie any sort of performance characteristics to a version of plv8 is really not advised, unless it is at a patch level, since v8 itself has drastically different performance characteristics across its versions.

I think it's disingenuous to pretend that moving that from V8 to a bytecode interpreted JS engine is not going to have *wildly* different and unexpected performance profile than a typical major version upgrade would have, not to mention the potential instability of other JS engines.

> and you just made a bunch of assumptions - a) that a new package was never an option

I'm trying to make it clear that it's the only option available if you're concerned about users and cloud providers dropping support as it seemed to me to be the premise of opening this discussion.

> if you spent time looking at the lack of commonality between javascript engines, you likely would not suggest this - there is less commonality between javascript engines than between different languages

I've spent the time and it's not an insurmountable problem at all. ChakraCore[1], Graal[2] and YODAOS[3] were all able to implement N-API so targeting a subset of N-API is one potential avenue. Mozilla's work on WASI[4] and Wasmer's `wasm` PG extension[5] could also be a source of inspiration, because plv8's subset of types could easily be represented across the WASM call boundary.

Anyway, I've got a branch that uses the Node.js build pipeline that alleviates all the concerns other than stable ABI issues: https://github.com/devinus/plv8/commit/b1f6d3f830f33c56b4e0b549944ab55f416da3c4

Some notes:

1. Using the Node.js pipeline to build V8 only requires gcc/g++/python/make[6]
2. Node.js packages are pretty ubiquitous and availably on almost every platform you're concerned about including Windows
3. The V8 version for Node.js 12.3.0 is also 7.4.288
4. Node.js takes care of ICU target/host endianness differences automagically
5. All tests pass

[1]: https://blogs.windows.com/msedgedev/2017/07/27/node-chakracore-update-n-api-ios/
[2]: https://hacks.mozilla.org/2019/03/standardizing-wasi-a-webassembly-system-interface/
[3]: https://github.com/graalvm/graaljs/blob/master/graal-nodejs/doc/api/n-api.md
[4]: https://github.com/yodaos-project/ShadowNode
[5]: https://github.com/wasmerio/postgres-ext-wasm
[6]: https://github.com/nodejs/node/blob/master/BUILDING.md#unix-prerequisites

**Comment by claytongulick** [comment]
>I've managed to work around most of those issues, but it would still not be a direct replacement without additional work.

Given how closely plv8 is tied to v8 (I did a cursory review of the source) that's pretty damn impressive that you already have a PoC up with QuickJS. Is it in a repo somewhere that I could look at? I'd like to learn more about the internals of PG language extensions.

>but currently it does not cleanly compile under all plv8 supported platforms out of the box, nor directly support all of the features that plv8 currently supports (even ignoring isolates at the moment). in order to be a full replacement, the code would need to be forked until (if?) the upstream supported all of the platforms/use cases. 

So what are your thoughts? Is switching to a different engine a moot point if there's not (practically) an engine that would work? Do you have a preferred choice? Have you seen XS7?

Over [here](https://news.ycombinator.com/item?id=20957677) someone mentioned XS7, they [seem to have good platform support](https://github.com/Moddable-OpenSource/moddable/tree/public/xs/platforms) and it was designed to be small footprint and embeddable.

Regarding performance characteristics, I'm not sure how much it really matters.

It seems to me like 99% of the use case of plv8 is simply making JSON manipulation easier to feed RESTful APIs, or to have node/js devs create complex db-side business rules with decreased cognitive burden of language switching - but all the heavy lifting is still done by pg via plv8.execute. 

It's difficult for me to believe that folks are writing CPU intensive javascript tasks that run in PG, and if they are, they may need to rethink their architecture - i.e. would anyone use pgplsql for image pixel manipulation?

Regardless of the direction you pick, I'm happy to help out. Up to my gills in client work right now, but I can find time to take some tasks.

**Comment by JerrySievert** [comment]
> I think it's disingenuous to pretend that moving that from V8 to a bytecode interpreted JS engine is not going to have _wildly_ different and unexpected performance profile than a typical major version upgrade would have, not to mention the potential instability of other JS engines.

and nowhere did I say that, I simply pointed out the types of performance changes, and under what use cases they would appear.  you should note that changing to the "v8 lite" model, which might be the case with some users/providers given v8's large memory footprint, also adapts this model which changes performance as well.

> I'm trying to make it clear that it's the only option available if you're concerned about users and cloud providers dropping support as it seemed to me to be the premise of opening this discussion.

that does appear to be your opinion, yes.  you're attempting to speak with full authority for all of those use cases, and aren't presenting evidence to back that up.  this discussion was opened in response to dwindling support and cloud providers actively dropping support, and your insistence is that the only way that anyone will move forward is to stay the course.  your opinion has been heard.

> Anyway, I've got a branch that uses the Node.js build pipeline that alleviates all the concerns other than stable ABI issues: devinus@b1f6d3f

I've said it before, and I'll say it again, pull requests are welcome.  but, it would still need the buy-off of the package maintainers to be considered, and would need to successfully work on all platforms that plv8 supports, this includes not just compiling on windows but working through any g++/msvc interoperability challenges.  given that most distributions do not support node 12, its might be a tough sell, but the discussion can be had.

I will finish by noting that the commonality between node and plv8 v8 versions is not an accident.

**Comment by JerrySievert** [comment]
@claytongulick 

> Given how closely plv8 is tied to v8 (I did a cursory review of the source) that's pretty damn impressive that you already have a PoC up with QuickJS. Is it in a repo somewhere that I could look at? I'd like to learn more about the internals of PG language extensions.

not really that impressive, QuickJS's api is pretty straightforward, and the PoC ignores a few things:

* caching - contexts should be cached at a user id level, as well as pointers to functions.  quickjs's startup time is fast enough to make creating a new context every time less important, but for a release it needs to be addressed
* memory contexts - conversions and execution should live in separate memory contexts, so long use of this PoC in a session will likely leak memory
* SPI - calls into postgres are simply code and type coercion, but in the interest of determining feasibility I chose to skip past them for the time being

you're welcome to take a look, I've created a repo at https://github.com/JerrySievert/pljs but given the things I explicitly ignored you shouldn't use it as a starting point for your own language.  a better repo to look at to learn from would be @RhodiumToad's Lua extension: https://github.com/RhodiumToad/pllua-ng as they've made sure to address those requirements where my PoC was simply to test viability.

> It seems to me like 99% of the use case of plv8 is simply making JSON manipulation easier to feed RESTful APIs, or to have node/js devs create complex db-side business rules with decreased cognitive burden of language switching - but all the heavy lifting is still done by pg via plv8.execute.

I've spent a lot of time both inside users enterprise use cases, and inside obscure cloud provider test cases.  These typically do not occur via issue, as they are usually sensitive - I tend to either get an email or on occasion a phone call, and work through any issues gratis.  For the most part, you are correct in regards to the bulk of usage that I've seen: JSON manipulation.  You can see any changes that this has prompted in the plv8 code regarding where new memory contexts are used or in the `JSONB_DIRECT_CONVERSION` code.  But, there are some specific use-cases leveraging javascript's module ecosystem mostly focused on crypto.   This is where most CPU focused use-cases would be affected, and would need to be fleshed out fully.  These specific use-cases would likely best be solved with a native extension but the developers using them are mostly tied to dealing with _trusted_ mainstream extensions allowed by cloud providers.

I do want to pause for a moment and make sure there is an understanding of _trusted_ in this context.  _Trusted_ in postgres language extension nomenclature refers to an extension that sandboxes the language and does not allow access to outside facilities such as system calls.  It has nothing to do with how someone "feels" about the language or environment, it's a designation for extensions that do not allow access to resources outside of the database itself.  For instance, `pl/python` and `pl/perl` both contain both _trusted_ and _untrusted_ versions and can change behavior based on how they are instantiated.

**Comment by mitar** [comment]
I must echo some of voices in this thread that one of main reasons why I like this project is because it is using v8. It is simply the most secure, fast engine, and feature complete out there. In fact, because of v8 JavaScript itself is so quick that it is faster than any other popular interpreter language (like Python and stuff), because Google throws so much resources into it. This also makes it ugly. There is no performance improvements without ugly code. Optimizations are ugly.

If I understand correctly, the main reason behind this is because existing maintainers in Debian decided to stop maintaining plv8 Debian packages? Isn't then the simplest solution simply to search for new maintainers? It could even be somebody from plv8 community. I think it would be less work overall than a) maintaining a v8 fork and/or b) rewriting this package without v8. So anyone can maintain Debian packages. It is not some closed group of people. I have in the past simply maintain a package for my own project myself.

(I think that moving closer to node.js sounds like a good idea.)

**Comment by mitar** [comment]
BTW, have you checked out also [Duktape](https://duktape.org/)? I have used it pretty successfully in one other embedded JS project.

**Comment by JerrySievert** [comment]
I've used duktape in a few projects over the years. it's fairly slow, but what really takes it out of the running is the lack of features.  it has partial support for es2015 and es2016, but that's where it stops, leaving it a bit behind the pack.

that said, it's api is decent, and pretty stable.

**Comment by mitar** [comment]
Thanks. Glad to know.

**Comment by markokr** [comment]
If the switch to V8 sourced from NodeJS is so simple as above patch shows, shouldn't it be done on principle anyway, as noone is happy about current build process?

I don't think there is a need for any sort of "buy-in" from anybody, and how would that even look like?

I think if PL/V8 process would be described as:

* For building PL/V8 versions 5.x you need source tarball of NodeJS 16.x.

it would be much more attractive to distro maintainers.  Although probably distro people would prefer to target node LTS releases as they change less disruptively.  So PL/V8 could require either Node 10.x (current LTS) or be ready for next, Node 12.x (future LTS).

As for portability - it does not make sense that PL/V8, a smallish project, tries to take responsibility of building V8 everywhere that PostgreSQL builds.  Instead the tagline should be "PL/V8 supports systems where PostgreSQL and NodeJS both build in a compatible way."  And that's it.  The project should worry only about own code, not external build systems.

**Comment by pkit** [comment]
@markokr
> Node 10.x (current LTS) 

I'm not sure if using proven insecure release is a good fit for a trusted module.

**Comment by JerrySievert** [comment]
> I don't think there is a need for any sort of "buy-in" from anybody, and how would that even look like?

the primary reason this issue has been opened is due to plv8 being dropped from *all* linux packages, ranging from operating systems (debian, ubuntu, fedora, redhat), and by the official Postgres releases themselves.  this has caused large consumers of plv8 to drop support completely, with the *latest* being heroku dropping all plv8 support completely and removing it wholesale from its cloud offerings.

it might be worth noting that modern nodejs is also *not* supported by the operating system maintainers on these systems - in order to support nodejs you must install secondary repositories maintained by a third-party company, nodesource.  see the official nodejs documentation: https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions-enterprise-linux-fedora-and-snap-packages

> it would be much more attractive to distro maintainers. Although probably distro people would prefer to target node LTS releases as they change less disruptively.

except that distro maintainers do not support this model.  you state this without any authority, whereas the distro maintainers with full authority have already said the opposite, and repeated it recently - see https://github.com/plv8/plv8/issues/308#issuecomment-532110552 for the most recent comment, but before stating what distro maintainers want, it would be best to read that whole thread, which lays out exactly what distro maintainers want and do not want.

> So PL/V8 could require either Node 10.x (current LTS) or be ready for next, Node 12.x (future LTS).

nodejs 10.x is far behind plv8 in terms of v8 versions, and is not remotely compatible, does not have the same security updates, and does not have the same javascript features.  nodejs 12 is compatible, but still does not meet the requirements.

> As for portability - it does not make sense that PL/V8, a smallish project, tries to take responsibility of building V8 everywhere that PostgreSQL builds

except that it is plv8's mission to work where postgres and v8 are capable of working.

> Instead the tagline should be "PL/V8 supports systems where PostgreSQL and NodeJS both build in a compatible way." And that's it. 

plv8 is in no way related to nodejs, whereas it is directly related to postgres.  they serve very different, specific tasks.  it's like saying that a squash is the same thing as an apple, they're both fruit, but that's where the similarities start to fall apart.   you may as well say that chrome and nodejs are the same thing, since they both embed v8.

> The project should worry only about own code, not external build systems.

except v8, which is embedded as per the requirement of the v8 team.  it indeed is ridiculous to worry about external build systems that are not related to plv8, such as nodejs.

**Comment by markokr** [comment]
>> As for portability - it does not make sense that PL/V8, a smallish project, tries to take responsibility of building V8 everywhere that PostgreSQL builds

> except that it is plv8's mission to work where postgres and v8 are capable of working.

As these discussions have proven that V8 is capable of working - in a maintainable way - only inside Chrome, the misson needs to change?

>> The project should worry only about own code, not external build systems.

> except v8, which is embedded as per the requirement of the v8 team. it indeed is ridiculous to worry about external build systems that are not related to plv8, such as nodejs.

The V8 team requirement just means they do not care about providing usable package for such uses.

So PL/V8 has proven current "embedded V8" approach does not work.  And PL/V8 people claim that linking with NodeJS does not work.

Does that leave any other option than "switch JS engine"?

If the goal is portable PL/JS that runs in most places PostgreSQL runs at, V8 may indeed be wrong engine to use.

**Comment by mitar** [comment]
> except that distro maintainers do not support this model. 

I disagree with this characterization. The comment you linked in fact directly supports the idea of linking against libnode. So why not do that?

**Comment by mitar** [comment]
> it might be worth noting that modern nodejs is also not supported by the operating system maintainers on these systems

Not sure why you repeating this, if I already showed you that they are supported?

https://packages.debian.org/experimental/nodejs

**Comment by JerrySievert** [comment]
> Not sure why you repeating this, if I already showed you that they are supported?

https://packages.debian.org/experimental/amd64/libnode72/filelist

missing anything related to v8 7.4.  even minor versions of v8 are not guaranteed to be compatible with each other, v8 does not use semantic versioning.

also, having something be *experimental* is not the same as having it supported.  nor does pointing to a single experimental repository for debian mean that ubuntu, fedora, or redhat are supported.  your argument simply does not make sense.

it's also worth noting that nodejs compiles with different options and requirements than plv8 does.  this is by design, as plv8 is specifically a trusted extension.  having external startup data would defeat this.

**Comment by mitar** [comment]
> even minor versions of v8 are not guaranteed to be compatible with each other, v8 does not use semantic versioning.

Yes, you would have to build against this particular version of node. See comments above how to do this. It also looks it would simplify the building process.

> also, having something be experimental is not the same as having it supported.

This is just the name of the distribution. New packages enter into experimental, move to unstable, testing, then stable. See it at other versions:

https://packages.debian.org/source/sid/nodejs
https://packages.debian.org/source/bullseye/nodejs
https://packages.debian.org/source/buster/nodejs

So this is how Debian works. What this means is:
* That the package is maintained.
* And that new version of node are coming in. They might take time, but they are coming in.

So linking against node and then:
* Distribute plv8 against latest version of node by ourselves (and PostgreSQL) to get the latest.
* Debian and other distributions link against the version they package, which might be an older one. But still, you can package it.

> it's also worth noting that nodejs compiles with different options and requirements than plv8 does. this is by design, as plv8 is specifically a trusted extension. having external startup data would defeat this.

Hm, I do not know anything about this, but I doubt that extension could not be simply made to not use those additional "external" features of node, nor expose them to users.

**Comment by JerrySievert** [comment]
> Yes, you would have to build against this particular version of node. See comments above how to do this. It also looks it would simplify the building process.


the API changes, which means that plv8 source would need to change to support this *older* version of v8, which has more security issues, as well as less javascript features.  v8 versions are not compatible with each other, cannot be compiled against code that use other versions of v8, and will simply crash if attempted to load a version of the library it is not compatible with, as the ABI changes as well.

it's not a matter of "compiling against a different version", it's a matter of supporting very specific versions of v8.

> Distribute plv8 against latest version of node by ourselves (and PostgreSQL) to get the latest.

so now plv8 is not only in the plv8 package maintenance and distribution chain for all of these operating systems (see the previous matrix comment), but now distributing postgres as well.  that's a non-starter.  if you want to personally build and maintain package repositories for all of the supported operating systems, supported postgres versions on those operating systems, and the platforms/architectures that are supported, then I'm happy to point documentation there: the same way that the nodejs project does.  even nodejs, which has a *lot* of corporate money behind it, does not directly support that model, but instead points to the nodesource packages.

> Debian and other distributions link against the version they package, which might be an older one. But still, you can package it.

this just points out that you do not understand v8: it's not technically possible.  you cannot simply swap out a binary version of v8, the ABI is not compatible, it has never been, and likely never will be.  you also cannot simply swap out and build against a different version of v8.  the API is not compatible, it has never been, and likely never will be.  this is not a plv8 thing, this is a v8 thing.  instantiating a v8 runtime changes from version to version, converting types changes from version to version, isolates change from version to version, memory management changes from version to version, executing javascript changes from version to version.  all of this is at the c++ level for compilation and at the ABI level for execution.

> Hm, I do not know anything about this, but I doubt that extension could not be simply made to not use those additional "external" features of node, nor expose them to users.

yes, that occurs at *compile time*.

I encourage you to spend some time looking at v8, its configuration, its API changes from version to version, and think about those implications.  forget that nodejs even exists for a minute, as they (like plv8) completely abstract that away from you as a consumer, and look at the v8 project directly - here's a good place to start: https://docs.google.com/document/d/1g8JFi8T_oAE_7uAri7Njtig7fKaPDfotU6huOa1alds/edit (that's the API changes from version to version).  pay close attention to the changes between 7.2 and 7.4, which includes the removal of all non-maybe versions of the API - which is an extremely breaking change between versions.

**Comment by eliaspanagiotidis** [comment]
Reading the entire above conversation, in my opinion there is no way to avoid abandoning V8. However, the bigger advantage of plv8 is its speed. Replacing V8 with an interpreted JS engine, such as QuickJS (which has been used in Jerry's PoC), most probably will adversely affect this speed to a significant extent (let alone credibility). It would be perfect if a well supported fast embeddable engine with JIT compiler (such like ChakraCore or SpiderMonkey) could take the place of V8.

**Comment by JamesInform** [comment]
I have often heard that Spidermonkey is faster than V8. When you look at the current browsers I think Firefox does really have the edge on performance when it comes to executing arbitrary complex JavaScript code.

But the question is, what will happen to existing JavaScript code written for the current PLV8 extension. Will a change to e.g. Spidermonkey require to rewrite big parts of the source code or will the existing JavaScript mainly work?

From my point of view it‘s also important not to sacrifice speed. Otherwise we will run into the issue that the performance of existing systems that rely on PLV8 is degraded.

Another question: If I understand correctly, there will be no ES2016 compatibility in PLV8 (as long as it is based on V8), right?

**Comment by eliaspanagiotidis** [comment]
Both SpiderMonkey and ChakraCore are comparable with V8 in terms of speed and their compatibility with modern versions of ES is pretty similar.
In my opinion, an adoption of either engine does not enforce any changes to the PLV8's API and existing javascript code's compatibility is guaranteed, except from some real extreme cases.
Even the change in the object's name (PLJS instead of PLV8) could be easily addressed by setting an alias in the first versions of the PLJS (or even by the user in the start_proc), so that existing code does not need to be modified at all.

**Comment by jakob** [comment]
I'm late to the discussion, but as the maintainer of Postgres.app I would just like to say that I would absolutely love it if the project abandoned v8. Compiling v8 takes about 50% of the time of compiling Postgres.app, and it's been a major source of build issues. I shipped outdated versions of plv8 for years because I couldn't figure out how to build newer versions. I think at some point we had to drop support for older versions of macOS because v8 didn't build on them any more.

I've spent many hours dealing with v8 build issues in the past.

Currently the process works (thanks a lot to the person that wrote the code that automatically builds v8), but as @JerrySievert says, wo knows if it'll work on the next version of macOS.

So from a maintainer perspective, your work on finding a replacement for v8 is greatly appreciated.

**Comment by claytongulick** [comment]
Not trying to add noise, but has deno been considered?

One of the stated goals of the project is easy embedding, and they have a pretty interesting security first approach to code execution. 

They haven't hit 1.0 yet, but are getting there pretty quick, it seems.

**Comment by tracker1** [comment]
@claytongulick came in to ask the same... I'm not sure if the build/embed for deno is at a better stability point, but might be a cleaner integration given deno's approach for embedding.  Of course, I haven't tried to build deno, and it would still have more limited targets than PosgreSQL itself, but at the very least could be provided where there is support, just x64 linux, windows, macOS would be the majority of coverage... though some arm and mips targets longer term would probably be a good idea, but I don't think it's nearly as urgent.

**Comment by JerrySievert** [comment]
unfortunately, deno doesn't help with a main issue: distribution.  it just adds another barrier that package maintainers won't deal with - another unsupported (by them) package required to build a library that they can't distribute as a shared library.

**Comment by syrusakbary** [comment]
I think Deno's approach might be a great fit. Otherwise [Wasmer Postgres extension](https://github.com/wasmerio/postgres-ext-wasm) could really be a great contender :)

**Comment by waaeer** [comment]
There is one more JS engine: https://github.com/nginx/njs

**Comment by Todd-Fulton** [comment]
Just a hobbyist here, but I wouldn't mind at all if the engine was switched. I've been trying to figure out how to build this on arm64 (ubuntu20.04 on raspberry pi) for about half a day now and the biggest problem is that v8 doesn't support building natively on arm, and their build process is not something I'm familiar with to just go in and edit a few make or cmake files, it's a bunch of python and custom generators. Will probably take me a few days to figure out the flow of it all and if I can hack it up to build natively on my pi's.

From this experience I can totally understand why any sane package manager would refuse to touch v8 with a 10' pole, and by extension pl/v8.

Sorry I don't have advice on how to accomplish such a task, but I do agree it might be a good move to switch if you want to support a wider array of platforms and keep the support of the various linux distro package managers.

**Comment by mariusa** [comment]
Newbie comment: When looking at alternatives, most will look better because the issues with the current one are known, whereas alternatives only present the good parts until one goes into depth, maybe in months (eg https://bugs.chromium.org/p/v8/issues/detail?id=9368). 

As @pkit pointed out, V8 is used by nodejs across various platforms and is offered by many hosting companies, including Heroku, which dropped plv8 support. Being a newbie, I don't understand how V8 is good enough for nodejs, but not for plv8. 

Also, V8 is the JS engine inside Chromium, which has big companies developing it: Google & Microsoft. Theoretically, these should fix performance & security bugs better than other small JS engines.

My plea would be to reconsider keeping V8 and thinking more about what would be needed to make it viable and even included in the standard PostgreSQL distribution at some point.

**Comment by emrul** [comment]
I’m really not an expert on this stuff but I want to ask if you’ve looked into GraalVM https://www.graalvm.org/ which would allow embedding JavaScript as well as a bunch of other languages under one runtime. Also embedded in Oracle’s database.

**Comment by pkit** [comment]
@emrul Oracle and GPL? No, thanks.

**Comment by emrul** [comment]
> @emrul Oracle and GPL? No, thanks.

GPL 2 with Classpath exception so should be fine shouldn’t it?

**Comment by eliaspanagiotidis** [comment]
@emrul GPL 2 is not fine since it still enforces the same license to all software using the module. And Oracle's behaviour with Java and MySQL reveals the future of this project also.

**Comment by tracker1** [comment]
@eliaspanagiotidis @emrul beyond all of this, given the Google/Android patent dispute, Oracle does not operate in good faith and cannot be trusted with anything less than very permissive licensing.

**Comment by JerrySievert** [comment]
kind of a moot point anyways, gpl would pretty much mean it's a non-starter.

**Comment by ppKrauss** [comment]
Hi,  this project [seems **not** abandoned](https://github.com/plv8/plv8/graphs/contributors?from=2019-01-27&to=2020-06-16&type=c) and is very **important** (!), but it is not easy to install, seems not serious or not reliable for UBUNTU users like me... There are some work on "plug and play" installation? How to help?

**Comment by emrul** [comment]
I use it on Ubuntu no issues here. Suggest asking for help on stackoverflow (this thread is about V8)

**Comment by dfcsoftware** [comment]
Having spent several days trying to build plv8 on arm71 (never worked) and X86_64 (worked after excellent help from @pkit and @JerrySievert to remove apt package libnode-dev), I started re-writing hundreds of lines of plv8 into pl/pgsql. Since PG enhanced json/jsonb operators on v_10/11/12 I feel comfortable moving off plv8. As mentioned by @claytongulick, json support was my main driver moving into plv8 five years ago. 

Now the support is too much as my project needs to work on Windows, MacOS, Linux and armX SBC like Raspberry-Pi and BeagleBone. I've already shed mod_perl for the same support issues. I'm building a car, not the road.

I will sorely miss plv8 and hope this group can come up with solutions. It's a beautiful road.

**Comment by pkit** [comment]
Revisiting the Cloudflare Workers discussion after a year, when they did launch a free plan.
It seems like they are using Chrome/Chromium instrumentation and launch code directly inside a browser using a ServiceWorker interface (pure guess, but the lack of debugging tools...)
I suppose some sort of `SecureWorker` interface will arise pretty soon and then Postgres or any other software that wants to execute code can do it that way.
But I think it's kind of a full architecture overhaul.

**Comment by tracker1** [comment]
I know it's been a while since mentioned... but SpiderMonkey/libmozjs may be the best path forward if a broadly supported JS engine with good embedding story are desired.  Not to mention ESR (extended support release) library versions in various distros.

https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/How_to_embed_the_JavaScript_engine

**Comment by disarticulate** [comment]
I think, without reading _everything_, the happy path would be one with some how can integrate with the webpack or rollup or other bundler build paths.

In that scenario, it would allow access to the whole node.js/browser environment. I understand this is coming from the other end of the spectrum, given that v8 is the low level library, with node on top or chrome on top.

(i'm writing this while building the whole v8; slowly)

looking over: https://en.wikipedia.org/wiki/Comparison_of_JavaScript_engines

Looks like Gecko is the most full featured, and is whats running with Firefox 90.

Obviously if there's a rewrite going on, finding a way to harness the power of the extensive browser based libraries and providing browser like environment APIs (websockets? webrtc?) would be cool.

**Comment by JerrySievert** [comment]
if there were websockets, or webrtc, plv8 could no longer be a *trusted* extension.  sorry, but that's a non-starter.

**Comment by luss** [comment]
Plv8 is great and it’s trusted.   If someone wants plNodeJS as an I trusted
language, then that’s a different pl and may (or may not) be a great idea.
I personally love pl/Java and pl/python,  but because thry’re untrusted i
usually have to code in either plpgsql or plv8

**Comment by ppKrauss** [comment]
> Hi, this project [seems **not** abandoned](https://github.com/plv8/plv8/graphs/contributors?from=2019-01-27&to=2020-06-16&type=c) and is very **important** (!)

But, ok **let's try NodeJS**... Any suggestions, **is there a PL-NodeJS implementation?**

Is there a kind of [NodeJS isolated-vm](https://github.com/laverdet/isolated-vm) adapted for PostgreSQL? Does anyone here tested?

**Comment by disarticulate** [comment]
Deno might be security conscious enough for a backend

**Comment by JerrySievert** [comment]
there are things in the works.

please stand by.

**Comment by ppKrauss** [comment]
> Deno might be security conscious enough for a backend

Ok, ["option 2" (instead NodeJS's isolated-vm)](#issuecomment-1086998422) is **Deno**... how to use it with PostgreSQL?  You are using?

1. install as [github.com/denoland/deno](https://github.com/denoland/deno)   instructios, right? nothing else?
2. ... and... Where the code that implements it as a PostgreSQL extension? <br/> With *isolated-vm* we can **reuse** a lot of  *plv8* sources (is it?), but with Deno we need to translate it? 
3. ... test: good? No bugs?

**Comment by tracker1** [comment]
Regarding Deno.. since it's already mapped in Rust... could be a matter of a binary module using [pgx](https://github.com/tcdi/pgx) (used with ZomboDB project).

**Comment by pkit** [comment]
> Regarding Deno.. since it's already mapped in Rust... could be a matter of a binary module using [pgx](https://github.com/tcdi/pgx) (used with ZomboDB project).

Deno is pretty insecure:
```
Deno 1.20.5
exit using ctrl+d or close()
> Object.prototype[Symbol.iterator] = function() {
     return {
       next:() => this
     }
  };
  [...({})];

<--- Last few GCs --->

[80853:0x559d1ccdaaa0]     4019 ms: Scavenge 564.2 (597.1) -> 564.2 (597.1) MB, 52.4 / 0.0 ms  (average mu = 0.724, current mu = 0.724) allocation failure; 


<--- JS stacktrace --->


#
# Fatal javascript OOM in invalid array length
#

Trace/breakpoint trap (core dumped)
```

**Comment by JerrySievert** [comment]
deno is also not supported by any major repo, so it is a non-starter.

**Comment by kirkw** [comment]
Sharing a Users Perspective...
We are coming to PG from Oracle.  We have leveraged the AWS SCT, which, of course, leverages PLV8.
Kudos to you guys, there is a lot to love about PLV8...
But we come from the Windows World.  If you thought building this was hard on Linux... OMG...
We are days into this process, trying to get this compiled and working for PG14...
PG12 was so easy, xtuple made a package for windows.  Plopped it in.

Our goal was to get this compiled and offer it back to the community.  Hah, we are fighting with msvc issues,
Nina, CMake redefinition, etc. etc. etc.  We finally got V8.dll to compile (Python2 and Python3 swapping, we
added this to the list to try to fix the python2 code so that it all runs on Python3).

We are still exploring our options and extensions and performance issues.  With 70k lines of PLSQL to convert...
Most of that in packages with package variables.  This building PLV8 hiccup is making us re-think our approach!

Reading this thread actually doubles our concerns.  Also, the realization that this will be an Annual investment is not lost on me.

Again, we were hoping to give back to pay it forward in appreciation for all that came before us.  So, some costs are acceptable.  You are talking to my struggle.  Do I engineer the PLV8 out of my solution today?  If so, YOU should be aware (and probably are), that this does not bode well for future adoption.

With that said... What problem can we solve or help solve?  Your challenge is you are middleware glue between 2 masters (PG and V8).  You have to support both.  Now, in the Windows world, it would be acceptable to build the V8.dll and provide it with the PLV8.dll, etc.  But that doesn't fly in the Linux world where you build everything for that hardware, etc.  Plus, it's not your core focus (IMO) as everyone is mostly Linux based.

What is the smallest thing that could be done to reduce 80% of the build pain? [It may be different on the two platforms]
My list would be:
a) Tool Chain simplification [at least remove python2 as a dependency, may only be in windows]
b) Tool Chain Validation
c) Publish the Windows 64bit Extensions like xtuple did [Simplifying a build could be just eliminating it for many]
d) Decouple the V8 build  from the PLV8 build??? [which is what having a Package Manager handle it means, right?]

Please forgive my ignorance in general, but (d) in specific.  I find if I allow myself to be wrong out load, I learn faster, and others may learn something as well...

I have hired a guy to build this in Linux on a VM so I can play with it.  Also to report what he runs into, starting from scratch.
This may show that "some" of the issues are simply in the windows version.  (char *fields[size]; // size must be a constant),
redeclared stat structure, the Python2 issue may be...

Finally, "my list", who am I...  A nobody, a noob.  I am truly curious.  What would be the smallest thing that would have the biggest impact in simplification?

Also, what about the NUGET V8?
https://github.com/pmed/v8-nuget

Or does this open a can of worms of it's own?
Thanks in advance...

**Comment by JerrySievert** [comment]
thank you for your well written comment, but I just want to respond to one part:

> d) Decouple the V8 build from the PLV8 build

that's a decision on the v8 side, on how they have chosen to present and distribute v8, unfortunately not in the control of the plv8 team, hence this whole discussion.

if that were an easy fix, then believe me, it would have been done long ago.

**Comment by kirkw** [comment]
Thanks for the clarification.
Would it be possible, if not to fork the V8 Project and clean up the build issues...
To create an OVERLAY repository, named by V8 version, so that after we download V8, before we build, we hit this.  This creates a directory with the files we need to overlay before running "our" build process, that simplifies the V8 Build Process?

My Thought here is that this overlay would have to be updated if/when V8 is updated, but that's about the time the PLV8 is agreeing to support this new version (because of a potential of breaking changes).

Now, this is the "beginning" of an uneducated question.  But if your process says "I can link with V8 revision 123", then your part of the install grabs that specific revision, grabs the Overlay for building that Revision (and the Overlay might well be different for Windows vs Linux).  And then we proceed to build V8 Easier (because of the Overlay).

Then recompile PLV8 with the inputs.

[ie, every problem in computers is often solved by adding a layer of indirection...  I am trying to find something that will take a Windows Person a few hours and be able to share back, and a Linux Person a few hours and share back (for the Overlay), once the Overlay is in place.  I would gladly sponsor such a solution if it helps!  Realizing the initial work is 80% of the cost!]

Am I getting close to the problem? [Again, I only see the problem through my brand new glasses of V8 Build Pain before I can even see if PLV8 builds.  Maybe there is more I should understand...  Youngest of 4 boys... I take criticism REALLY well. LOL]

**Comment by JerrySievert** [comment]
> Would it be possible, if not to fork the V8 Project and clean up the build issues...

yes, but plv8 is maintained by a single volunteer maintainer.  to do this would take a lot of time and effort, plus be a continuous ongoing effort for any further v8 changes (there are breaking changes on minor version, keeping track of major versions and those commits would be a full time job on its own).

currently, nodejs has several of those maintainers.

v8 has a very complicated build system, that by default downloads its own compilers, linkers, all requirements to build, validates it all, and builds for you.  replicating that is very expensive and time consuming.

> Now, this is the "beginning" of an uneducated question. But if your process says "I can link with V8 revision 123"

version 123, with these specific compiler settings and compilation settings against v8, where a single small difference will crash.  so, no.  almost every project requires slightly different settings, for various reasons.

> I would gladly sponsor such a solution if it helps!

great!  we'd need 4-5 people pretty much full time for 3-4 months (experienced guess), to rebuild the build system, plus someone to manage that full-time.  that's my best off-the-cuff cost.  so, maybe $250k just to get going.  for me, moving to a better suited javascript engine is much cheaper and a better solution.

**Comment by luss** [comment]
Is there a better Javascript version u are actively considering?  I'm
having trouble building the google stuff in CentOS 7 these dayz.

**Comment by JerrySievert** [comment]
> Is there a better Javascript version u are actively considering? 

a large amount of work has been done against quickjs.  it has tradeoffs, but builds everywhere with very little effort.

**Comment by kirkw** [comment]
> great! we'd need 4-5 people pretty much full time for 3-4 months (experienced guess), to rebuild the build system, plus someone to manage that full-time. that's my best off-the-cuff cost. so, maybe $250k just to get going. for me, moving to a better suited javascript engine is much cheaper and a better solution.

Excellent Detail that cuts to the chase.  It's not a breadbox!  This is the problem in the end.  The scale of work on either choice.

So, since this is way beyond my ability to sponsor, either we hunt for a sponsor, or we review QuickJS as suggested.
I will go review building QuickJS on the Windows end (we may end up abandoning Windows b/c of this, or PLV8 which is sad).

What is the effort to switch to QuickJS if that builds?  (Again forgive my ignorance, as I first started this thread it was VERY Vague to me what the root issue was and the scale of the effort to fix it).  

So my assumptions of next steps are:
1) Review QuickJS, make sure it builds well [do we need to ASK for help with specific Linux, OS versions?]
2) [Hard Part] Rework PLV8 to link to this V8
3) Regression Testing/Debugging/Stability Testing
4) Redo the Build Process/Documentation, Include the "Run These Tests Before installing!"
5) Merge into the main branch

**Is there a short list of (We know QuickJS (QJS) will not work if/when).**  (Fail Early)

I will work on the Windows QJS, and have my resource build it on Linux (Ubuntu) and give me feedback.
If I can do more, please ask.  I have access to a PhD developer I can pay to work on reconnecting QJS to PLV8.
I am willing to pay a reasonable consulting fee to even setup some pair programming or knowledge transfer To you/your team.  My assumption being that once we reconnect the ABI/API points, and know how to test/validate it...  That this will transfer backwards into Linux for you (offering the skills I have access to).

[I will need a few days for some other priorities...]
Thanks for putting up with me. :-)

EDIT: Okay, I already have a libquickjs.dll files, prebuilt for windows.  That was easy.
200% on board with helping convert PLV8 to use QJS...

**Comment by JerrySievert** [comment]
@kirkw most of the steps you are outlining are complete, but this will not be merged into plv8, it will be its own project/repo under the plv8 organization.

I just have not had the time to push my current status up, and have not as it is very early alpha.

plv8 is extremely tightly coupled against v8, it is not designed nor architected to replace the engine, and carries a very large amount of v8 related baggage.  plv8 will continue as-is on its own, but future users should use the replacement where it makes sense.

**Comment by kirkw** [comment]
@JerrySievert You are a RockStar, thanks!  Okay, this is perfect clarification.
V8 will ultimately be replaced and we will probably end up with a PLJS Extension that's future ready. (Reserve the EXT name now, LOL).  It's not lost on me that your initials are JS, ROTFLMAO!

Sorry to have wasted your time.  But this post will save others!

My guy said QuickJS built on Ubuntu in 3 minutes.  Great Choice!

Finally, pg_background threads do not have the same PLV8JS context as the caller... (not a deal breaker, but an issue to be careful about for our migration), I assume this is similar to Parallel Query Issues.

FWIW: I sent you a few cups of coffee via the sponsor button. Thank you for taking the time out to answer my questions.  I gained, and you should too.  Thank you (and the team) for a wonderful package.

You can reach out if you need someone like me to test/validate especially on the Windows side!

**Comment by JerrySievert** [comment]
> Finally, pg_background threads do not have the same PLV8JS context as the caller.

correct, Postgres's model is not multi-threaded, it's multi-process.  so, sharing a context between processes becomes a lot harder.   safer and cleaner to just tie a context to a process.

**Comment by tracker1** [comment]
> > Regarding Deno.. since it's already mapped in Rust... could be a matter of a binary module using [pgx](https://github.com/tcdi/pgx) (used with ZomboDB project).
> 
> Deno is pretty insecure:
> 
> ```
> Deno 1.20.5
> exit using ctrl+d or close()
> > Object.prototype[Symbol.iterator] = function() {
>      return {
>        next:() => this
>      }
>   };
>   [...({})];
> ```

Just curious how this is handled currently with plv8?  A similar issue happens with Node.

**Comment by JerrySievert** [comment]
@tracker1 safely.

see https://github.com/plv8/plv8/blob/r3.1/sql/array_spread.sql

**Comment by tracker1** [comment]
@JerrySievert have you filed issues, respectively to Node or Deno on this?

**Comment by JerrySievert** [comment]
it is a very well-known v8 bug.

**Comment by pkit** [comment]
> @JerrySievert have you filed issues, respectively to Node or Deno on this?

It's a v8 problem, their OOM handling is atrocious and there are no plans to fix it. As it "works" for Chrome (they just use proccess isolation).
This one is just one problem of many.

**Comment by luss** [comment]
Hi Kirk,

Have you tried building the latest version of PLV8 with the MSYS64 GNU
Toolchain on Windows?    I am hopeful to get that solution working myself
in the near future.

--Luss

**Comment by luss** [comment]
Clearly a simple ubiquitous build process would get plv8 more traction.
What are the cons to QuickJS.   I just took a quick look at the project and
it looks like a part-time one man show.  I assume making a JavaScript
compiler from scratch is not an easy or quick thing to do.

**Comment by JerrySievert** [comment]
quickjs is extremely solid with a fairly active embedding community.  while it does not accept PR's, there is an active mailing list where patches are accepted.

it also builds cleanly on all platforms, and I have a working PoC that is not feature complete - just haven't had time to complete a MVP for initial release.

what I have left:

* arrays between javascript and Postgres
* queries from inside javascript (plv8.execute, etc)
* memory/process management

as for plv8 and its build process, I have trying to find someone to help further on expanding my build process changes, and to expand on the docker images.  I'm meeting with someone later today that I might be able to sponsor to continue further on the docker portion, and then possibly into the more generic build process.

**Comment by kirkw** [comment]
> Have you tried building the latest version of PLV8 with the MSYS64 GNU Toolchain on Windows?

Unfortunately we have decided against JS due to the build issues and the future uncertainty.  We switched from AWS-SCT to Ora2Pg.  The Risks were just too high for the team.

**Comment by serefarikan** [comment]
I've come across v8's build blues in various scenarios: trying to use a high performance javascript engine in one case or another. Frankly, it goes well over my head. I also came across QuickJs as an alternative, but [the performance difference was significant](https://github.com/bellard/quickjs/issues/79), so I kept sticking to wrappers around v8 created by people much smarter than me. I know synthetic benchmarks are not indicative of real life performance, but the perf. difference is hard to ignore, at least according to QuickJs's own benchmarks. It may be worth including performance as criteria for consideration.

**Comment by JerrySievert** [comment]
@serefarikan as you noted, those benchmarks are extremely synthetic.

and yes, all performance considerations between plv8 and pljs are being noted (and will be called out very loudly), but I did have one thing to share:

the startup difference between v8 and quickjs is fairly extreme.  for most stored procedures that do data manipulation, quickjs ends up being faster because of these differences.  for much longer running stored procedures (think 5+ seconds), v8 definitely wins out.

so, any performance differences need to be considered for how each stored proc behaves, and decisions for users/developers will need to be based on these differences.

plv8 will continue to exist, with improved building and will continue to be updated, but for 95% of scenarios, having a much easier to build and support tool would end up being better.

**Comment by serefarikan** [comment]
@JerrySievert thanks, that's a great approach. The startup difference is a very interesting factor to consider, didn't even occur to me, but now that you said it, it should have been an obvious thing to think about. Many thanks for your great work btw, before I forget :)

**Comment by pkit** [comment]
@JerrySievert there's a relatively new project https://github.com/bnoordhuis/v8-cmake that builds V8 completely independently of google pipelines, AFAIK.
Need to check it out.

**Comment by JerrySievert** [comment]
thanks for the pointer @pkit! (and nice to see you back).

I'll take a look at it.

in the mean time, there is also https://github.com/plv8/pljs - early alpha, but getting close to feature complete.  still need to benchmark the heck out of everything to show pros and cons.  it's important that everything be transparent as to when it makes sense to use what.

isolates and wasm will likely never make it into pljs.

**Comment by JerrySievert** [comment]
I would like to follow up on a few things:

1. there is now a 3.2alpha branch that changes the build system once again to [v8-cmake](https://github.com/bnoordhuis/v8-cmake), which also upgrades v8 to version 11.5.  there were a few PR's required to make it work, but they were accepted gracefully.
2. an issue was opened in April with concerns about performance regressions between versions of plv8 (and thus v8): https://github.com/plv8/plv8/issues/531 - I did a bunch of benchmarking (the benchmarks are available in the thread) around `plv8.execute`, which is specifically the v8 javascript/c++ membrane, I also did the comparison to pljs and the differences are huge and concerning.  unfortunately, this is due to underlying changes in v8.

**Comment by pkit** [comment]
I have switched to gvisor in my current projects, but AFAIR newer v8 were consistently faster for my use cases (mainly serving nodejs-like api/pages from PG) than older ones.
I may have time to run benches next month, but I do remember consistently outperforming nodejs in their own tests, lol.

**Comment by JerrySievert** [comment]
yeah, I'm not surprised about that, but specifically crossing the membrane is what appears to have slowed down.

**Comment by pkit** [comment]
V8 introduces more and more checks there for validity, so it probably makes sense. But the last numbers in your bench are huge, begs some research on what did they change.

**Comment by ppKrauss** [comment]
[*Bun runtime*](https://bun.sh/docs), is a viable alternative? Perhaps the [problems are similar to Deno](#issuecomment-577229123)...

https://github.com/oven-sh/bun

![image](https://github.com/plv8/plv8/assets/1651447/6940a75e-20a7-445a-a6c2-7ade24f6483d)

**Comment by pkit** [comment]
@ppKrauss Bun is a runtime. Plv8 is a runtime too. Running runtime inside runtime is...suboptimal.
Bun uses [JavaScriptCore](https://github.com/WebKit/WebKit/tree/main/Source/JavaScriptCore).
Plv8 uses [V8](https://v8.dev/).
So, the discussion could be: maybe use JavaScriptCore instead of V8? 
But Bun itself is useless here.

**Comment by ppKrauss** [comment]
> So, the discussion could be: maybe use JavaScriptCore instead of V8?

Oops, yes, it is.

**Maybe use [JavaScriptCore](https://github.com/WebKit/WebKit/tree/main/Source/JavaScriptCore) instead of V8?**

PS: [Bun](https://bun.sh/docs) proves that is a reliable process, and the Bun community can help understand the details.

**Comment by JerrySievert** [comment]
https://github.com/plv8/pljs

**Comment by JerrySievert** [comment]
closing due to pljs existing
